contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md

```diff
@@ -0,0 +1,340 @@
+
+### APIs
+
+**Scope**:
+
+| Union Type | Supported |
+|---|---|---|
+| non-inlined non-discriminated union | Yes |
+| non-inlined discriminated union | Yes |
+| inlined union with [patchMergeKey](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#strategic-merge-patch) only | Yes |
+| other inlined union | No |
+
+For the inlined union with patchMergeKey, we move the tag to the parent struct's instead of
+adding some logic to lookup the metadata in go struct of the inline union.
+Because the limitation of the latter is that the metadata associated with
+the inlined APIs will not be reflected in the OpenAPI schema.
+
+#### Tags
+
+old tags:
+
+1) `patchMergeKey`:
+It is the key to distinguish the entries in the list of non-primitive types. It must always be
+present to perform the merge on the list of non-primitive types, and will be preserved.
+
+2) `patchStrategy`:
+It indicates how to generate and merge a patch for lists. It could be `mrege` or `replace`. It is optional for lists.
+
+new tags:
+
+`patchStrategy: replaceKeys`:
+
+We introduce a new value `replaceKeys` for `patchStrategy`. It indicates the all the fields that
+need to be preserved must present in the patch. And it will be merged with live object.
+All the missing fields will be cleared when patching.
+
+#### Examples
+
+1) Non-inlined non-discriminated union:
+
+Type definition:
+```go
+type ContainerStatus struct {
+	...
+	// Add patchStrategy:"replaceKeys"`
+	State ContainerState `json:"state,omitempty" protobuf:"bytes,2,opt,name=state" patchStrategy:"replaceKeys"``
+	...
+}
+```
+Live object:
+```yaml
+state:
+  running:
+    startedAt: ...
+```
+Local file config:
+```yaml
+state:
+  terminated:
+    exitCode: 0
+    finishedAt: ...
+```
+Patch:
+```yaml
+state:
+  $patch: replaceKeys
+  terminated:
+    exitCode: 0
+    finishedAt: ...
+```
+Result after merging
+```yaml
+state:
+  terminated:
+    exitCode: 0
+    finishedAt: ...
+```
+
+2) Non-inlined discriminated union:
+
+Type definition:
+```go
+type DeploymentSpec struct {
+	...
+	// Add discriminator:"<discriminator-name>"
+	Strategy DeploymentStrategy `json:"strategy,omitempty" protobuf:"bytes,4,opt,name=strategy" discriminator:"type"`
+	...
+}
+```
+Since there are no fields associated with `recreate` in `DeploymentSpec`, I will use a generic example.
+
+Live object:
+```yaml
+unionName:
+  discriminatorName: foo
+  fooField:
+    fooSubfield: val1
+```
+Local file config:
+```yaml
+unionName:
+  discriminatorName: bar
+  barField:
+    barSubfield: val2
+```
+Patch:
+```yaml
+unionName:
+  $patch: replaceKeys
+  discriminatorName: bar
+  barField:
+    barSubfield: val2
+```
+Result after merging
+```yaml
+unionName:
+  discriminatorName: bar
+  barField:
+    barSubfield: val2
+```
+
+3) Inlined union with `patchMergeKey` only.
+This case is special, because `Volumes` already has a tag `patchStrategy:"merge"`.
+We change the tag to `patchStrategy:"merge|replaceKeys"`
```

Comment:
`merge` means we should merge the list.
`replaceKeys` mean we should replace keys for each entry in the list,